sw/device/tests/dif/dif_alert_handler_unittest.cc

```diff
@@ -0,0 +1,922 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_alert_handler.h"
+
+#include <cstring>
+#include <limits>
+#include <ostream>
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "alert_handler_regs.h"  // Generated.
+
+namespace dif_alert_handler_unittest {
+namespace {
+using ::mock_mmio::LeInt;
+using ::mock_mmio::MmioTest;
+using ::mock_mmio::MockDevice;
+using ::testing::_;
+using ::testing::Return;
+
+constexpr uint32_t kAlerts = 12;
+constexpr uint32_t kAllOnes = std::numeric_limits<uint32_t>::max();
+
+class AlertTest : public testing::Test, public MmioTest {
+ protected:
+  dif_alert_handler_t handler_ = {.params = {
+                                      .base_addr = dev().region(),
+                                      .alert_count = kAlerts,
+                                      .escalation_signal_count = 4,
+                                  }};
+};
+
+class InitTest : public AlertTest,
+                 public testing::WithParamInterface<uint32_t> {};
+
+TEST_P(InitTest, Success) {
+  dif_alert_handler_params_t params = {
+      .base_addr = dev().region(),
+      .alert_count = GetParam(),
+      .escalation_signal_count = 4,
+  };
+
+  dif_alert_handler_t handler;
+  EXPECT_EQ(dif_alert_handler_init(params, &handler), kDifAlertHandlerOk);
+}
+
+TEST_P(InitTest, NullArgs) {
+  dif_alert_handler_params_t params = {
+      .base_addr = dev().region(),
+      .alert_count = GetParam(),
+      .escalation_signal_count = 4,
+  };
+
+  EXPECT_EQ(dif_alert_handler_init(params, nullptr), kDifAlertHandlerBadArg);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    InitTestSignalCounts, InitTest,
+    testing::Values(1, 2, 12, 16
+                    // TODO: Enable these parameters once we can support them.
+                    // See: #3826
+                    // 24, 32,
+                    ));
+
+class ConfigTest : public AlertTest {
+  // We provide our own dev_ member variable in this fixture, in order to
+  // support IgnoreMmioCalls().
+  //
+  // NOTE: This must come before handler_!
+ private:
+  std::unique_ptr<MockDevice> dev_ =
+      std::make_unique<testing::StrictMock<MockDevice>>();
+
+ protected:
+  ConfigTest() { handler_.params.base_addr = dev().region(); }
+  // Non-virtual-override dev(), so that EXPECT_*() functions look this one
+  // up instead of AlertTest::dev().
+  MockDevice &dev() { return *dev_; }
+
+  // Disables expectations on MMIO operations. This is useful for configuration
+  // tests that partially configure the device but fail due to a configuration
+  // error, returning kError.
+  void IgnoreMmioCalls() {
+    dev_ = std::make_unique<testing::NiceMock<MockDevice>>();
+    handler_.params.base_addr = dev().region();
+
+    // Make sure that the peripheral looks unlocked.
+    ON_CALL(*dev_, Read32(_)).WillByDefault(Return(kAllOnes));
```

Comment:
This made me wish that `InSequence` in `MmioTest` was opt-out. I gave `Times(AnyNumber())` a try (while keeping the mock strict) but it didn't work.

Kudos for not writing MMIO expectations for calls that return `k...Error`.

Reply:
I think it might not be unreasonable to generalize your un-sequenced mixin!